peer/peer.go

```diff
@@ -0,0 +1,1942 @@
+	p.StatsMtx.Lock()
+	defer p.StatsMtx.Unlock()
+
+	log.Tracef("Updating last blk for peer %v, %v", p.addr, blkSha)
+	p.lastAnnouncedBlock = blkSha
+}
+
+// AddKnownInventory adds the passed inventory to the cache of known inventory
+// for the peer.  It is safe for concurrent access.
+func (p *Peer) AddKnownInventory(invVect *wire.InvVect) {
+	p.knownInvMutex.Lock()
+	defer p.knownInvMutex.Unlock()
+
+	p.knownInventory.Add(invVect)
+}
+
+// Id returns the peer id.
+func (p *Peer) Id() int32 {
+	return p.id
+}
+
+// NA returns the peer network address.
+func (p *Peer) NA() *wire.NetAddress {
+	return p.na
+}
+
+// Addr returns the peer address.
+func (p *Peer) Addr() string {
+	return p.addr
+}
+
+// Filter returns the peer bloom filter.
+func (p *Peer) Filter() *bloom.Filter {
+	return p.filter
+}
+
+// Inbound returns whether the peer is inbound.
+func (p *Peer) Inbound() bool {
+	return p.inbound
+}
+
+// Persistent returns whether the peer is persistent.
+func (p *Peer) Persistent() bool {
+	return p.persistent
+}
+
+// RetryCount returns the number of times to retry peer connection.
+func (p *Peer) RetryCount() int64 {
+	return p.retryCount
+}
+
+// RequestQueue returns the peer request queue.
+func (p *Peer) RequestQueue() []*wire.InvVect {
+	return p.requestQueue
+}
+
+// SetRequestQueue sets the peer request queue to the given queue.
+func (p *Peer) SetRequestQueue(requestQueue []*wire.InvVect) {
+	p.requestQueue = requestQueue
+}
+
+// QueueRequest queues a peer request.
+func (p *Peer) QueueRequest(iv *wire.InvVect) {
+	p.requestQueue = append(p.requestQueue, iv)
+}
+
+// LastAnnouncedBlock returns the last announced block of the peer.
+func (p *Peer) LastAnnouncedBlock() *wire.ShaHash {
+	p.StatsMtx.Lock()
+	defer p.StatsMtx.Unlock()
+
+	return p.lastAnnouncedBlock
+}
+
+// LastPingNonce returns the last ping nonce of the peer.
+func (p *Peer) LastPingNonce() uint64 {
+	p.StatsMtx.Lock()
+	defer p.StatsMtx.Unlock()
+
+	return p.lastPingNonce
+}
+
+// LastPingTime returns the last ping time of the peer.
+func (p *Peer) LastPingTime() time.Time {
+	p.StatsMtx.Lock()
+	defer p.StatsMtx.Unlock()
+
+	return p.lastPingTime
+}
+
+// LastPingMicros returns the last ping micros of the peer.
+func (p *Peer) LastPingMicros() int64 {
+	p.StatsMtx.Lock()
+	defer p.StatsMtx.Unlock()
+
+	return p.lastPingMicros
+}
+
+// VersionKnown returns the whether or not the version of a peer is known
+// locally.  It is safe for concurrent access.
+func (p *Peer) VersionKnown() bool {
+	p.StatsMtx.Lock()
+	defer p.StatsMtx.Unlock()
+
+	return p.versionKnown
+}
+
+// ProtocolVersion returns the peer protocol version in a manner that is safe
+// for concurrent access.
+func (p *Peer) ProtocolVersion() uint32 {
+	p.StatsMtx.Lock()
+	defer p.StatsMtx.Unlock()
+
+	return p.protocolVersion
+}
+
+// ContinueHash returns the continue hash.
+// It is safe for concurrent access.
+func (p *Peer) ContinueHash() *wire.ShaHash {
```

Comment:
The continue hash is specific to the full block downloading mechanism and isn't directly related to the peer.  This is another piece of data that should be tracked by the block manager.

Given these various pieces of information, I suspect it might make sense to create another struct in btcd to house additional peer data and have a single `map[int64]*peerData` on the block manager which houses it all.